### DISCUSSION TITLE
The attribute zoom-min="2" behaves like zoom-min="1" in the xml file with rendertheme

### DISCUSSION THREAD

**Discussion by TomasJohansson** [mapsforge/vtm#931] [Q&A] [answered]
The problem/behaviour I am describing below is the same for my two devices, which is one tablet with Android 4.4
and one phone with Android 9.

I have a labeling problem in an app where I am using 'mapsforge vtm'.
The problem is that I am trying to start showing labels at zoom level 2, i.e. I do not want the labels to be shown at level 1,
but those labels are still shown at level 1.

Therefore I have now created a very small minimal app that reproduces the problem, and I have attached it in a zip file below.
[map1.zip](https://github.com/mapsforge/vtm/files/8787854/map1.zip)

I am using Windows, and within the above attached zip file there is a bat file "map_creation\create_map.bat" which can be run (assuming that 'osmosis' and mapsforge mapfile-writer is installed).
When the bat file is executed, it generates a file "map.map" which is copied into the directory "app\src\main\assets"
(but I have also left such a generated map file there in the attached zip file within the assets folder, so you do not need to generate it yourself)

When the app is started, the map file (and the xml file with rendertheme) is automatically copied from the assets folder into the android app's directory which is then used as 'MapFileTileSource'.

The bat file is creating the map file with this command:
```
`osmosis --rx file=map.osm --mapfile-writer file=map.map tag-conf-file=./tag-mapping.xml zoom-interval-conf=2,1,3 tag-values=true
```
Below is the most relevant (regarding this question/issue) content from the file "map_creation\map.osm":
```
    <node id='19' lat='48.0' lon='-99.0' version="1" timestamp="2022-05-27T14:20:23Z" >
        <tag k="name" v="North America" />
        <tag k="zoom_level" v="1_5" />
    </node>
    <node id='29' lat='-13.0' lon='-58.0' version="1" timestamp="2022-05-27T14:20:23Z" >
        <tag k="name" v="South America" />
        <tag k="zoom_level" v="2_5" />
    </node>
    <node id='39' lat='9.0' lon='20.0' version="1" timestamp="2022-05-27T14:20:23Z" >
        <tag k="name" v="Africa" />
        <tag k="zoom_level" v="3_5" />
    </node>    
```

Below is the most relevant content from the file "map_creation\tag-mapping.xml":
```
    <pois>
        <osm-tag key="zoom_level" value="1_5" />
        <osm-tag key="zoom_level" value="2_5" />
        <osm-tag key="zoom_level" value="3_5" />
    </pois>
```

Below is the most relevant content from the file "app\src\main\assets\mapsforge_render.xml":
```
    <m e="node" k="zoom_level">

        <m v="1_5" zoom-min="1" zoom-max="5">
            <text use="style_continent_name" />
        </m>

        <m v="2_5" zoom-min="2" zoom-max="5">
            <text use="style_continent_name" />
        </m>

        <m v="3_5" zoom-min="3" zoom-max="5">
            <text use="style_continent_name" />
        </m>

    </m>
```

The map file (generated from "map_creation\map.osm") defines three polygons
(with only three nodes i.e. triangles) which *very* roughly corresponds to three continents.
My intention (with the above configuration in the file "mapsforge_render.xml")
is that the labels should be displayed at different minimal zoom levels (zoom-min 1, 2, 3).

It works as expected when zooming in from level 2 to level 3, i.e. then the label "Africa" becomes shown, and it becomes hidden again when zooming out to level 2 again.
(see the first attached screenshot below for zoom level 3 shown in a 'Toast' message).
However, when I have zoomed to level 1, then the label "South America" is still shown, although my intention is that it should only be shown at levels 2+
(see the second attached screenshot below for zoom level 1 shown in a 'Toast' message).

In other words, the problem shown in the second screenshot below (at zoom level 1) is that only the label for 'North America' should be displayed, but *not* the label for 'South America' which should only be shown for level 2 and above.
![zoom_level_3](https://user-images.githubusercontent.com/1504507/170735101-e73ecd0b-1c53-4738-8ce6-7f280443491c.gif)
![zoom_level_1](https://user-images.githubusercontent.com/1504507/170735161-bd963ff4-46cd-44eb-845f-a351910d70e1.gif)

**Comment by devemux86** [comment, marked as answer]
- The `ViewPort` setters must be called before creating the layers

- The layer constructors with the **custom** zoom levels are needed:
```java
MapFileTileSource mapFileTileSource = new MapFileTileSource(1, 5);
map.layers().add(new LabelLayer(map, tileLayer, new LabelTileLoaderHook(), 5));
```

**Comment by TomasJohansson** [comment]
Yes indeed it worked. Thank you !
